api-server/stack-test-suite/tests/in_memory.rs

```diff
@@ -0,0 +1,1160 @@
+};
+use hex::ToHex;
+use rstest::rstest;
+use serde_json::json;
+use std::{net::TcpListener, sync::Arc};
+use test_utils::random::{make_seedable_rng, Rng, Seed};
+use web_server::{api::web_server, APIServerWebServerState};
+
+#[tokio::test]
+async fn server_status() {
+    let url = "/";
+
+    let listener = TcpListener::bind("127.0.0.1:0").unwrap();
+    let socket = listener.local_addr().unwrap();
+
+    tokio::spawn(async move {
+        let web_server_state = {
+            let chain_config = Arc::new(create_unit_test_config());
+            let storage = TransactionalApiServerInMemoryStorage::new(&chain_config);
+
+            APIServerWebServerState {
+                db: Arc::new(storage),
+                chain_config: Arc::clone(&chain_config),
+            }
+        };
+
+        web_server(listener, web_server_state).await
+    });
+
+    let response = reqwest::get(format!("http://{}:{}{url}", socket.ip(), socket.port()))
+        .await
+        .unwrap();
+
+    assert_eq!(response.status(), 200);
+    assert_eq!(response.text().await.unwrap(), r#"{"versions":["1.0.0"]}"#);
+}
+
+#[tokio::test]
+async fn bad_request() {
+    let url = "/non-existent-url";
+
+    let listener = TcpListener::bind("127.0.0.1:0").unwrap();
+    let socket = listener.local_addr().unwrap();
+
+    tokio::spawn(async move {
+        let web_server_state = {
+            let chain_config = Arc::new(create_unit_test_config());
+            let storage = TransactionalApiServerInMemoryStorage::new(&chain_config);
+
+            APIServerWebServerState {
+                db: Arc::new(storage),
+                chain_config: Arc::clone(&chain_config),
+            }
+        };
+
+        web_server(listener, web_server_state).await
+    });
+
+    let response = reqwest::get(format!("http://{}:{}{url}", socket.ip(), socket.port()))
+        .await
+        .unwrap();
+
+    assert_eq!(response.status(), 400);
+    assert_eq!(response.text().await.unwrap(), r#"{"error":"Bad request"}"#);
+}
+
+mod v1_block {
+    use super::*;
+
+    #[tokio::test]
+    async fn invalid_block_id() {
+        let url = "/api/v1/block/invalid-block-id";
+
+        let listener = TcpListener::bind("127.0.0.1:0").unwrap();
+        let socket = listener.local_addr().unwrap();
+
+        tokio::spawn({
+            async move {
+                let web_server_state = {
+                    let chain_config = Arc::new(create_unit_test_config());
+                    let storage = TransactionalApiServerInMemoryStorage::new(&chain_config);
+
+                    APIServerWebServerState {
+                        db: Arc::new(storage),
+                        chain_config: Arc::clone(&chain_config),
+                    }
+                };
+
+                web_server(listener, web_server_state).await
+            }
+        });
+
+        let response = reqwest::get(format!("http://{}:{}{url}", socket.ip(), socket.port()))
+            .await
+            .unwrap();
```

Comment:
I'm not too sure if this is actually a race...

The listening port is bound before splitting tasks, and so even if the socket hasn't been passed to `web_sever()` yet and so the `reqwest::get()` is called, since the socket is already listening, the `get()` will just block until it receives data.

I verified this on Mac OS by putting a `sleep(5)` on the first line within the `async move` task, and tests passed. Would also assume this works the same on all platforms.

Reply:
I don't think a sleep proves the race doesn't exist, because that depends on what kind of sleep you're doing and what. What kind of sleep did you use? If you used a synchronous sleep, that will just make the main async thread block and won't give the chance to other threads to take over, meaning that you probably didn't trigger proper concurrency. More importantly, you're using `tokio::test`, which is single-threaded, so you don't have parallelism anyway. 

To prove this, you should increase the number of threads and/or use async sleep.